### PR TITLE
Filter rows with null keys when coalescing due to reaching cuDF row limits [databricks]

### PR DESCRIPTION
**Pull request by abellina**
This PR is a workaround to handle cases where we would go above the cuDF row limit when coalescing a build-side batch in the hash join, but we have a chance to potentially rescue things if the batch is built mostly of nulls. The real fix is to not materialize those rows with null keys (https://issues.apache.org/jira/browse/SPARK-39131), but that fix in the logical plan introduces other issues that we don't have an answer to yet.

Posting this as a draft to get some comments on the implementation. Specifically on the join types that this applies to, it would be nice to get another pair of eyes there. 

I'll look into adding a test for this.

### CODE REVIEW

**Comment by abellina** [comment]
Ok, I may need to cover one more case, and that is where we still have `GpuCoalesceBatches(GpuShuffleCoalesce(shuffle))` for the build side. I am seeing this can still show up in the plan, even though we are trying to replace these actively via `shuffledHashJoinOptimizeShuffle`.

I am looking into it. Before I PRed this I had also changed `childrenCoalesceGoal` in `GpuShuffledHashJoinExec` to use `RequireSingleBatchWithFilter` instead of `RequireSingleBatch`, which totally works. But I am confused why the plan looks the way it looks. The change to `childrenCoalesceGoal` may require changes to other pattern matching, or to the inheritance structure so that `RequireSingleBatchWithFilter` and `RequireSingleBatch` have a common trait.

**Comment by abellina** [comment]
Ok, the reason for what I am seeing with some `GpuShuffledHashJoinExec` still having a build-side with `GpuCoalesceBatches(GpuShuffleCoalesce(shuffle))` is because the rule I added here (https://github.com/NVIDIA/spark-rapids/pull/4588) is too restrictive: https://github.com/NVIDIA/spark-rapids/blob/branch-22.06/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuTransitionOverrides.scala#L247, checking also that the left side has a coalesce batches.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCoalesceBatches.scala [hunk 1]
```diff
@@ -351,7 +373,14 @@ abstract class AbstractGpuCoalesceIterator(
 
       // there is a hard limit of 2^31 rows
       while (numRows < Int.MaxValue && !hasOnDeck && iter.hasNext) {
-        closeOnExcept(iter.next()) { cb =>
+        closeOnExcept(iter.next()) { cbFromIter =>
```
[the remark this answers is not in the document]
Comment (revans2) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCoalesceBatches.scala]: I think the problem is GpuFilter and us relying on it to close things for us. We should have a version that does not close and then we can have clearly defined boundaries.

[the remark this answers is not in the document]
Comment (abellina) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCoalesceBatches.scala]: yes that makes sense. Will do

[the remark this answers is not in the document]
Comment (abellina) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCoalesceBatches.scala]: I tried this, but at the end really in this case we always filter + close except for this last case I was left with, which seems like the simplest solution without refactoring more code is to null out the batch after it gets consumed by the filter: https://github.com/NVIDIA/spark-rapids/pull/5531/commits/74e16c2b94b27cfdb584c210ad398e42af0a1204. Let me know if you disagree @revans2

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCoalesceBatches.scala [hunk 2]
```diff
@@ -351,7 +373,14 @@ abstract class AbstractGpuCoalesceIterator(
...
+          val cb = if (inputFilterExpression.isDefined) {
+            // If we have reached the cuDF limit once, proactively filter batches
+            // after that first limit is reached.
+            GpuFilter.apply(cbFromIter, inputFilterExpression.get)
```
[the remark this answers is not in the document]
Comment (abellina) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCoalesceBatches.scala]: this should be done.

**Comment by abellina** [comment]
build

**Review by revans2 — APPROVED** [review]

**Comment by abellina** [comment]
triggering the databricks CI

**Comment by abellina** [comment]
build

**Review by jlowe — APPROVED** [review]

**Comment by abellina** [comment]
build

**Pull request merged**